File: Group-Video/OpenVideoCall-iOS-Objective-C/OpenVideoCall/LastmileViewController.m
//
//  LastmileViewController.m
//  OpenVideoCall
//
//  Created by 3 on 2020/12/15.
//  Portions Copyright (c) 2020 meta-rti. All rights reserved.
//

#import "LastmileViewController.h"

/// LastmileViewController implements various last mille related delegate functions for MetaRtcEngine.
/// "Last mile" refers to the connection between the local device and Meta's edge server.
@interface LastmileViewController () <MetaRtcEngineDelegate>

@property (weak, nonatomic) IBOutlet UILabel *qualityLabel;
@property (weak, nonatomic) IBOutlet UILabel *rttLabel;
@property (weak, nonatomic) IBOutlet UILabel *uplinkLabel;
@property (weak, nonatomic) IBOutlet UILabel *downlinkLabel;

@property (strong, nonatomic) UIActivityIndicatorView *activityView;
@property (assign, nonatomic) BOOL isLastmileProbeTesting;
@property (weak, nonatomic) MetaRtcEngineKit *metaKit;
@end

@implementation LastmileViewController
#pragma mark - Getter, Setter
- (MetaRtcEngineKit *)metaKit {
    return [self.dataSource lastmileVCNeedMetaKit];
}

- (void)setIsLastmileProbeTesting:(BOOL)isLastmileProbeTesting {
    if (isLastmileProbeTesting) {
        MetaLastmileProbeConfig *config = [[MetaLastmileProbeConfig alloc] init];
        config.probeUplink = YES;
        config.probeDownlink = YES;
        config.expectedUplinkBitrate = 5000;
        config.expectedDownlinkBitrate = 5000;
        [self.metaKit startLastmileProbeTest:config];
        [self.activityView startAnimating];
        self.title = @"Testing...";
    } else {
        [self.activityView stopAnimating];
        [self.metaKit stopLastmileProbeTest];
        self.title = @"Test result";
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self addActivityView];
    self.metaKit.delegate = self;
    self.metaKit.delegate = self;
    self.isLastmileProbeTesting = YES;
    self.view.backgroundColor = [UIColor whiteColor];
}

- (void)dealloc {
    self.isLastmileProbeTesting = NO;
}

#pragma mark - Private
- (void)addActivityView {
    UIActivityIndicatorView *activityView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhite];
    UIBarButtonItem *rightItem = [[UIBarButtonItem alloc] initWithCustomView:activityView];
    self.navigationItem.rightBarButtonItem = rightItem;
    self.activityView = activityView;
}

#pragma mark - MetaRtcEngineDelegate

/// Reports the last mile network quality of the local user once every two seconds before the user joins a channel.
/// @param engine - RTC engine instance
/// @param quality -The last mile network quality based on the uplink and dowlink packet loss rate and jitter. See list at:
- (void)rtcEngine:(MetaRtcEngineKit *)engine lastmileQuality:(MetaNetworkQuality)quality {
    NSString *string;
    switch (quality) {
        case MetaNetworkQualityExcellent:      string = @"excellent";   break;
        case MetaNetworkQualityGood:           string = @"good";        break;
        case MetaNetworkQualityPoor:           string = @"poor";        break;
        case MetaNetworkQualityBad:            string = @"bad";         break;
        case MetaNetworkQualityVBad:           string = @"very bad";    break;
        case MetaNetworkQualityDown:           string = @"down";        break;
        case MetaNetworkQualityUnknown:        string = @"unknown";     break;
        case MetaNetworkQualityDetecting:      string = @"detecting";   break;
        case MetaNetworkQualityUnsupported:    string = @"unsupported"; break;
    }
    self.qualityLabel.text = string;
}

/// Reports the last-mile network probe result.
/// @param engine - RTC engine instance
/// @param result - The uplink and downlink last-mile network probe test result, see list here:
- (void)rtcEngine:(MetaRtcEngineKit *)engine lastmileProbeTestResult:(MetaLastmileProbeResult *)result {
    self.rttLabel.text = [NSString stringWithFormat:@"%lu", (unsigned long)result.rtt];
    self.uplinkLabel.text = [NSString stringWithFormat:@"%lu", (unsigned long)result.uplinkReport.packetLossRate];
    self.downlinkLabel.text = [NSString stringWithFormat:@"%lu", (unsigned long)result.downlinkReport.packetLossRate];
    self.isLastmileProbeTesting = NO;
}
@end

File: Group-Video/OpenVideoCall-iOS/OpenVideoCall/LastmileViewController.swift
//
//  LastmileViewController.swift
//  OpenVideoCall
//
//  Created by 3 on 2020/12/16.
//  Portions Copyright (c) 2020 meta-rti. All rights reserved.
//

import UIKit
import MetaRTCFramework

protocol LastmileVCDataSource: NSObjectProtocol {
    func lastmileVCNeedMetaKit() -> MetaRtcEngineKit
}

class LastmileViewController: UITableViewController {
    
    @IBOutlet weak var qualityLabel: UILabel!
    @IBOutlet weak var rttLabel: UILabel!
    @IBOutlet weak var uplinkLabel: UILabel!
    @IBOutlet weak var downlinkLabel: UILabel!
    
    private var metaKit: MetaRtcEngineKit {
        return dataSource!.lastmileVCNeedMetaKit()
    }
    
    private var isLastmileProbeTesting = false {
        didSet {
            if isLastmileProbeTesting {
                let config = MetaLastmileProbeConfig()
                config.probeUplink = true
                config.probeDownlink = true
                config.expectedUplinkBitrate = 5000
                config.expectedDownlinkBitrate = 5000
                metaKit.startLastmileProbeTest(config)
                activityView?.startAnimating()
                self.title = "Testing..."
            } else {
                metaKit.stopLastmileProbeTest()
                activityView?.stopAnimating()
                self.title = "Test result"
            }
        }
    }
    
    private var activityView: UIActivityIndicatorView?
    
    weak var dataSource: LastmileVCDataSource?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        addActivityView()
        metaKit.delegate = self
        isLastmileProbeTesting = true
        view.backgroundColor = UIColor.white
    }
       
    deinit {
        isLastmileProbeTesting = false
    }
}

extension LastmileViewController: MetaRtcEngineDelegate {
    /// Reports the last mile network quality of the local user once every two seconds before the user joins a channel.
    /// - Parameters:
    ///   - engine: the Portions Copyright (c) 2020 meta-rti. All rights reserved. engine
    ///   - quality: An enum describing the network quality. Possible values are: Unknown = 0, Excellent = 1, Good = 2, Poor = 3, Bad = 4, VBad = 5, Down = 6, Unsupported = 7, Detecting = 8.
    func rtcEngine(_ engine: MetaRtcEngineKit, lastmileQuality quality: MetaNetworkQuality) {
        qualityLabel.text = quality.description()
    }
    
    /// Reports the last-mile network probe result.
    func rtcEngine(_ engine: MetaRtcEngineKit, lastmileProbeTest result: MetaLastmileProbeResult) {
        rttLabel.text = "\(result.rtt) ms"
        uplinkLabel.text = result.uplinkReport.description()
        downlinkLabel.text = result.downlinkReport.description()
        isLastmileProbeTesting = false
    }
}

extension LastmileViewController {
    func addActivityView() {
        let activityView = UIActivityIndicatorView(style: .white)
        let rightItem = UIBarButtonItem(customView: activityView)
        navigationItem.rightBarButtonItem = rightItem
        self.activityView = activityView
    }
}

extension MetaLastmileProbeOneWayResult {
    func description() -> String {
        return "\(packetLossRate)"
    }
}
